agent/src/discovery_handler_manager/discovery_handler_registry.rs

```diff
@@ -169,35 +171,60 @@ pub trait DiscoveryHandlerRegistry: Sync + Send {
 /// Real world implementation of the Discovery Handler Request
 struct DHRequestImpl {
     endpoints: RwLock<Vec<watch::Receiver<Vec<Arc<DiscoveredDevice>>>>>,
-    notifier: watch::Sender<Vec<Arc<DiscoveredDevice>>>,
+    notifier: watch::Sender<crate::device_manager::cdi::Kind>,
     key: String,
     handler_name: String,
     details: String,
     properties: Vec<DiscoveryProperty>,
-    extra_device_properties: HashMap<String, String>,
+    extra_device_properties: RwLock<HashMap<String, String>>,
     kube_client: Arc<dyn DiscoveryManagerKubeInterface>,
     termination_notifier: Arc<Notify>,
 }
 
+#[async_trait]
 impl DiscoveryHandlerRequest for DHRequestImpl {
-    fn get_instances(&self) -> Result<Vec<Instance>, DiscoveryError> {
+    async fn get_instances(&self) -> Result<Vec<Instance>, DiscoveryError> {
+        let properties = self.extra_device_properties.read().await;
         Ok(self
-            .notifier
-            .borrow()
+            .endpoints
+            .read()
+            .await
             .iter()
-            .map(|i| self.device_to_instance(i))
+            .flat_map(|r| r.borrow().clone().into_iter())
+            .map(|i| self.device_to_instance(i.as_ref(), &properties))
             .collect())
     }
+
+    async fn set_extra_device_properties(&self, extra_device_properties: HashMap<String, String>) {
+        let mut current = self.extra_device_properties.write().await;
+        if extra_device_properties != *current {
+            let edit = extra_device_properties
+                .iter()
+                .map(|(k, v)| format!("{}={}", k, v))
+                .collect();
+            *current = extra_device_properties;
+            self.notifier
+                .send_modify(|k| k.container_edits.first_mut().unwrap().env = edit);
+        }
+    }
 }
 
 impl DHRequestImpl {
-    fn device_to_instance(&self, dev: &DiscoveredDevice) -> Instance {
+    fn device_to_instance(
+        &self,
+        dev: &DiscoveredDevice,
+        extra_device_properties: &HashMap<String, String>,
```

Comment:
Why are you passing in properties instead of reading them from `self.extra_device_properties`?